[wip] Modify the dynamic flow control and add it to CFD-DEM (#746)
lethe-cfd/lethe: doc/source/parameters/cfd/dynamic_flow_control.rst

* The ``inlet boundary id`` parameter is the boundary where the average velocity intended is controlled (for CFD solvers only).

* The ``flow direction`` parameter sets the absolute direction. It should be ``0`` if the flow is in the :math:`x` direction, ``1`` in the :math:`y` direction and ``2`` in the :math:`z` direction.

* The ``initial beta`` parameter sets an initial :math:`\beta` coefficient that may speed up the convergence of the average velocity target. Some tests should be done to find a good initial :math:`\beta` coefficient.

.. tip:: 

  A good method to find a reasonable initial beta is to test two or three different initial beta parameters, write down the given flow rate at the first time step in the simulation and do a regression. The correlation is linear and giving a proper value will reduce the time it takes to reach the target velocity.

* The ``alpha`` parameter is a relaxation coefficient that is used to control the convergence speed or stability. The higher the value, the faster the convergence. However, if the value is too high, there might be oscillations and the simulation could be unstable.

> * The ``beta threshold`` parameter is the threshold on beta calculated at the previous time step that prevents the new calculated beta force to be updated, it is kept as the previous beta value. This could avoid the reassembly of the matrix because of the updated force term when reuse matrix option is enable for the non-linear solver

Reviewer: ```suggestion
* The ``beta threshold`` parameter is the threshold on beta calculated at the previous time step that prevents the new calculated beta force to be updated. If ``abs(beta_n - beta_n+1) < abs(beta_n * beta_threshold)``, the previous beta value is kept. This could avoid the reassembly of the matrix because of the updated force term when reuse matrix option is enabled for the non-linear solver.
```

.. note::

  If you are simulating a problem with periodic boundary conditions, it is important to control the average velocity by specifying the boundary id that corresponds to the inlet only.

* The ``inlet boundary id`` parameter is the boundary where the average velocity intended is controlled (for CFD solvers only).

* The ``flow direction`` parameter sets the absolute direction. It should be ``0`` if the flow is in the :math:`x` direction, ``1`` in the :math:`y` direction and ``2`` in the :math:`z` direction.

* The ``initial beta`` parameter sets an initial :math:`\beta` coefficient that may speed up the convergence of the average velocity target. Some tests should be done to find a good initial :math:`\beta` coefficient.

.. tip:: 

  A good method to find a reasonable initial beta is to test two or three different initial beta parameters, write down the given flow rate at the first time step in the simulation and do a regression. The correlation is linear and giving a proper value will reduce the time it takes to reach the target velocity.

> * The ``alpha`` parameter is a relaxation coefficient that is used to control the convergence speed or stability. The higher the value, the faster the convergence. However, if the value is too high, the convergence may be unstable.

Reviewer: ```suggestion
* The ``alpha`` parameter is a relaxation coefficient that is used to control the convergence speed or stability. The higher the value, the faster the convergence. However, if the value is too high, there might be oscillations and the simulation could be unstable.
```

> * The ``beta threshold`` parameter is the threshold of the previous calculated beta force where the new calculated beta doesn't need to be updated. This prevents the reassembly of the matrix because of the force term when reuse matrix is used for the non-linear solver

Reviewer: I don't really understand this explanation. The beta threshold is the old beta, or is it a tolerance in the sense that if abs(old_beta-new_beta)<tolerance, we reuse the same matrix?

Author: `beta_threshold` is a threshold on the `beta_n` value (for instance 0.01 is 1%). The way it works is that `beta_n+1` is calculated as usual but if `abs(beta_n - beta_n+1) < abs(beta_n * beta_threshold)`, it changes the value of `beta_n+1` to `beta_n`. That prevent to update the source term (the beta force is added to force in scratch data I think). It doesn't means the same matrix is going to be reuse with the option enable (this is independent of the non-linear solver parameters), but it could help to do so since we force the code to keep the same value. I will try to reformulate this explaination

Author: I reformulated the doc if you want to look at it :)

Reviewer: I understand better with your explanation, thanks. I think the inequality should be there as it really helps to understand (added as suggestion)

> * The ``beta threshold`` parameter is the threshold of the previous calculated beta force where the new calculated beta doesn't need to be updated. This prevents the reassembly of the matrix because of the force term when reuse matrix is used for the non-linear solver

Author: ```suggestion
* The ``beta threshold`` parameter is the threshold on beta calculated at the previous time step that prevents the new calculated beta force to be updated, it is kept as the previous beta value. This could avoid the reassembly of the matrix because of the updated force term when reuse matrix option is enable for the non-linear solver
```

Dynamic flow control
~~~~~~~~~~~~~~~~~~~~

The purpose of this subsection is to enable dynamic flow control. It is important when we want to simulate an average
velocity on a specific boundary (CFD) or the whole domain (for CFD-DEM). To control the average velocity of the flow, the code
calculates a :math:`\beta` coefficient at each time step that is used to keep the average velocity in an acceptable

> threshold. If the chosen wall boundary on which a average velocity is imposed is an inlet, the average velocity

Reviewer 2: ```suggestion
threshold. If the boundary on which an average velocity is imposed is an inlet, the average velocity
```

> targeted must be negative since the outward normal vector is used for the calculation. Thereby, if the chosen wall
> boundary is an outlet, average velocity must be positive. In other to keep the sign consistency, the targeted
> velocity should have an negative sign when used with CFD-DEM even if no normal vector is used since the
> whole domain is taken into account.
> 

Reviewer 2: ```suggestion
targeted must be negative since the outward normal vector is used for the calculation. If the 
boundary is an outlet, average velocity must be positive. In other to keep the sign consistency, the targeted
velocity should have an negative sign when used with CFD-DEM even if no normal vector is used since the
whole domain is taken into account.

```
These are boundaries, they are not wall (otherwise nothing would flow through them)

Author: Fair enough :p

> The main controller of the average velocity is the following equation and is based on Wang [1]:

Reviewer 2: ```suggestion
The main controller of the average velocity is the following equation and is based on approach of Wang [1]:
```

Reviewer 2: et al? are there many authors to the Wang paper?

Author: This is his/her thesis

Reviewer 2: great

.. math::
    \beta^{n+1} = \beta^n - \frac{\alpha}{\Delta t} \left[ (\bar{U})^{0} - 2(\bar{U})^{n} + (\bar{U})^{n-1} \right]

The default parameters are:

.. code-block:: text

  subsection flow control
    set enable                  = false
    set enable beta particle    = false
    set average velocity target = 0
    set boundary id             = 0
    set flow direction          = 0
    set initial beta            = 0
    set alpha                   = 1.0
    set verbosity               = quiet
  end

* The ``enable`` parameter is set to ``true`` to enable an imposed average velocity.

* The ``enable beta particle`` parameter is set to ``true`` to apply a proportional beta force on particles.

* The ``average velocity target`` parameter specifies the average velocity (:math:`m/s`). The value is compared to the calculated value at a boundary (for CFD solvers) or the whole domain (for CFD-DEM solver).

.. warning::

  This value should be negative if imposed on an the inlet boundary (or CFD-DEM), and positive if imposed on an outlet boundary.

.. note::

>   If you are simulating a problem with periodic boundary conditions at inlet and outlet walls, it is important to control the average velocity specifying the boundary id that corresponds to the inlet only (where the flow would have a negative value).

Reviewer 2: ```suggestion
  If you are simulating a problem with periodic boundary conditions, it is important to control the average velocity by specifying the boundary id that corresponds to the inlet only (where the flow would have a negative value).
```

Would it work if. you specify the outlet only? Why the inlet? isnt it just one of the boundary only and that's it?

Author: I don't remember very well why there's this comment. I think it would be better to use inlet only, so remove any mention of outlet boundary. Also, maybe I can remove the - sign for the normal vector, it makes more sense with the average velocity.

> * The ``boundary id`` parameter is the wall boundary where the average velocity intended is controlled (for CFD solvers only).

Reviewer 2: ```suggestion
* The ``boundary id`` parameter is the boundary where the average velocity intended is controlled (for CFD solvers only).
```
Ain't a wall though :)

* The ``flow direction`` parameter sets the absolute direction. It should be ``0`` if the flow is in the :math:`x` direction, ``1`` in the :math:`y` direction and ``2`` in the :math:`z` direction.

* The ``initial beta`` parameter sets an initial :math:`\beta` coefficient that may speed up the convergence of the average velocity target. Some tests should be done to find a good initial :math:`\beta` coefficient.

.. tip:: 

>   A good method to find a reasonable initial beta is to test two or three different initial beta parameters, write down the given flow rate at the first time step in the simulation and do a regression. The correlation is linear and giving a proper value will greatly speed up the convergence.

Reviewer 2: ```suggestion
  A good method to find a reasonable initial beta is to test two or three different initial beta parameters, write down the given flow rate at the first time step in the simulation and do a regression. The correlation is linear and giving a proper value will reduce the time it takes to reach the target velocity.
```
